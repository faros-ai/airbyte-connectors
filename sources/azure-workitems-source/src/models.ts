export interface WorkItemResponse {
  count: number;
  value: WorkItem[];
}

export interface WorkItem {
  fields: fields;
  id: string;
  rev: string;
  url: string;
}

export interface System {
  AreaPath: string;
  AssignedTo: user;
  BoardColumn: string;
  ChangedBy: user;
  ChangedDate: string;
  CreatedDate: string;

  IterationLevel3: string;
  IterationPath: string;
  PersonId: string;
  Reason: string;
  Rev: string;
  RevisedDate: string;
  State: string;
  TeamProject: string;
  Title: string;
  Watermark: string;
  WorkItemType: string;
  parent: string | null;
  Description: string;
}

export interface user {
  displayName: string;
  url: string;
  _links: string;
  id: string;
  uniqueName: string;
}

export interface User {
  subjectKind: string;
  domain: string;
  principalName: string;
  mailAddress: string;
  origin: string;
  originId: string;
  displayName: string;
  url: string;
  descriptor: string;
  _links: UserLink;
}

interface UserLink {
  self: Href;
  memberships: Href;
  membershipState: Href;
  storageKey: Href;
  avatar: Href;
}
interface Href {
  href: string;
}

export interface UserResponse {
  count: number;
  value: User[];
}

export interface fields {
  Microsoft: {
    VSTS: {
      Common: {
        Priority: string;
        StateChangeDate: string;
        ValueArea: string;
      };
    };
  };
  System: System;
}

export interface Iteration {
  attributes: {
    startDate: string;
    finishDate: string;
    timeFrame: string;
  };
  id: string;
  name: string;
  path: string;
  url: string;
}

export interface Board {
  id: string;
  name: string;
  url: string;
}
