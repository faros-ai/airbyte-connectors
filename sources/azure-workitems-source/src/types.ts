import {AzureDevOpsConfig} from 'faros-airbyte-common/azure-devops';

export interface AzureWorkitemsConfig extends AzureDevOpsConfig {
  readonly additional_fields: string[];
}
