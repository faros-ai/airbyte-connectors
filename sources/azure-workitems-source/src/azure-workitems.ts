import axios, {AxiosInstance, AxiosResponse} from 'axios';
import {wrapApiError} from 'faros-airbyte-cdk';
import {chunk, flatten} from 'lodash';
import {VError} from 'verror';

import {User, UserResponse, WorkItemResponse} from './models';
const DEFAULT_API_VERSION = '7.0';
const DEFAULT_GRAPH_VERSION = '7.1-preview.1';
const MAX_BATCH_SIZE = 200;
export const DEFAULT_REQUEST_TIMEOUT = 60000;

export interface AzureWorkitemsConfig {
  readonly access_token: string;
  readonly organization: string;
  readonly project: string;
  readonly api_version?: string;
  readonly request_timeout?: number;
  readonly graph_version?: string;
}

export class AzureWorkitems {
  private static azure_Workitems: AzureWorkitems = null;

  constructor(
    private readonly httpClient: AxiosInstance,
    private readonly graphClient: AxiosInstance
  ) {}

  static async instance(config: AzureWorkitemsConfig): Promise<AzureWorkitems> {
    if (AzureWorkitems.azure_Workitems) return AzureWorkitems.azure_Workitems;

    if (!config.access_token) {
      throw new VError('access_token must not be an empty string');
    }

    if (!config.organization) {
      throw new VError('organization must not be an empty string');
    }

    if (!config.project) {
      throw new VError('project must not be an empty string');
    }

    const version = config.api_version ?? DEFAULT_API_VERSION;
    const httpClient = axios.create({
      baseURL: `https://dev.azure.com/${config.organization}/${config.project}/_apis`,
      timeout: config.request_timeout ?? DEFAULT_REQUEST_TIMEOUT,
      maxContentLength: Infinity, //default is 2000 bytes
      params: {
        'api-version': version,
      },
      headers: {
        Authorization: `Basic ${config.access_token}`,
      },
    });

    const graphClient = axios.create({
      baseURL: `https://vssps.dev.azure.com/${config.organization}/_apis/graph`,
      timeout: config.request_timeout ?? DEFAULT_REQUEST_TIMEOUT,
      maxContentLength: Infinity,
      maxBodyLength: Infinity,
      params: {'api-version': config.graph_version ?? DEFAULT_GRAPH_VERSION},
      headers: {Authorization: `Basic ${config.access_token}`},
    });

    AzureWorkitems.azure_Workitems = new AzureWorkitems(
      httpClient,
      graphClient
    );
    return AzureWorkitems.azure_Workitems;
  }

  async checkConnection(): Promise<void> {
    try {
      const iter = this.getStories();
      await iter.next();
    } catch (err: any) {
      let errorMessage = 'Please verify your access token is correct. Error: ';
      if (err.error_code || err.error_info) {
        errorMessage += `${err.error_code}: ${err.error_info}`;
        throw new VError(errorMessage);
      }
      try {
        errorMessage += err.message ?? err.statusText ?? wrapApiError(err);
      } catch (wrapError: any) {
        errorMessage += wrapError.message;
      }
      throw new VError(errorMessage);
    }
  }

  private get<T = any, R = AxiosResponse<T>>(
    path: string
  ): Promise<R | undefined> {
    return this.handleNotFound<T, R>(() => this.httpClient.get<T, R>(path));
  }

  private async handleNotFound<T = any, R = AxiosResponse<T>>(
    call: () => Promise<R>
  ): Promise<R | undefined> {
    try {
      const res = await call();
      return res;
    } catch (err: any) {
      if (err?.response?.status === 404) {
        return undefined;
      }
      throw err;
    }
  }
  private post<T = any, R = AxiosResponse<T>>(
    path: string,
    data: any
  ): Promise<R | undefined> {
    return this.handleNotFound<T, R>(() =>
      this.httpClient.post<T, R>(path, data)
    );
  }

  async *getStories(): AsyncGenerator<any> {
    const data = {
      query:
        "Select [System.Id], [System.Title], [System.State] From WorkItems Where [System.WorkItemType] = 'User Story' order by [id] asc",
    };
    const list = await this.post<any>('wit/wiql', data);
    const ids: string[] = [];
    for (let i = 0; i < list.data.workItems.length; i++) {
      ids.push(list.data.workItems[i].id);
    }
    const ids2: string[] = [];
    const userStories = [];
    for (const id of ids) {
      if (ids2.length == MAX_BATCH_SIZE) {
        userStories.push(
          await this.get<WorkItemResponse>(
            `wit/workitems?ids=${ids2}&$expand=all`
          )
        );
        ids2.splice(0);
      }
      ids2.push(id);
    }
    if (ids2.length > 0) {
      userStories.push(
        await this.get<WorkItemResponse>(
          `wit/workitems?ids=${ids2}&$expand=all`
        )
      );
    }
    for (const item of userStories) {
      yield item;
    }
  }

  async *getWorkitems(): AsyncGenerator<any> {
    const promises = [
      "'Task'",
      "'User Story'",
      "'BUG'",
      "'Feature'",
      "'Epic'",
      "'Issue'",
      "'Product Backlog Item'",
      "'Requirement'",
    ].map((n) => this.getIdsFromAWorkItemType(n));

    const results = await Promise.all(promises);
    const ids: ReadonlyArray<string> = flatten(results);

    for (const c of chunk(ids, MAX_BATCH_SIZE)) {
      const res = await this.get<WorkItemResponse>(
        `wit/workitems?ids=${c}&$expand=all`
      );
      for (const item of res?.data?.value ?? []) {
        yield item;
      }
    }
  }

  async getIdsFromAWorkItemType(
    workItemsType: string
  ): Promise<ReadonlyArray<string>> {
    const data = {
      query:
        'Select [System.Id] From WorkItems WHERE [System.WorkItemType] = ' +
        workItemsType +
        ' order by [id] asc',
    };
    const list = await this.post<any>('wit/wiql', data);
    const ids = [];
    for (let i = 0; i < list.data.workItems.length; i++) {
      ids.push(list.data.workItems[i].id);
    }
    return ids;
  }

  async *getUsers(): AsyncGenerator<User> {
    let continuationToken: string;
    do {
      const res = await this.graphClient.get<UserResponse>('users', {
        params: {subjectTypes: 'msa,aad,imp', continuationToken},
      });
      continuationToken = res.headers?.['X-MS-ContinuationToken'];
      for (const item of res.data?.value ?? []) {
        yield item;
      }
    } while (continuationToken);
  }

  async *getIterations(): AsyncGenerator<any> {
    const res = await this.get<any>('work/teamsettings/iterations');
    let response;
    let response2;
    for (const item of res.data?.value ?? []) {
      response = await this.httpClient.get(item?.url);
      response2 = await this.httpClient.get(
        response?.data?._links?.classificationNode.href
      );
      if (typeof response2?.data?.id !== 'undefined') {
        item.id = response2.data?.id;
      }
      yield item;
    }
  }

  async *getBoards(): AsyncGenerator<any> {
    const res = await this.get<any>('work/boards');
    for (const item of res.data?.value ?? []) {
      yield item;
    }
  }
}
