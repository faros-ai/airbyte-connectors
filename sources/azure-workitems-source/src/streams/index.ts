import {Boards} from './boards';
import {Iterations} from './iterations';
import {Users} from './users';
import {Workitems} from './workitems';

export {Workitems, Users, Iterations, Boards};
