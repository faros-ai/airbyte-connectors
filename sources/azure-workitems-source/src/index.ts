import {Command} from 'commander';
import {
  AirbyteLogger,
  AirbyteSourceBase,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import VError from 'verror';

import {AzureWorkitems, AzureWorkitemsConfig} from './azure-workitems';
import {Boards, Iterations, Users} from './streams';
import {Workitems} from './streams/workitems';

/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteLogger();
  const source = new AzureWorkitemsSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** Example source implementation. */
export class AzureWorkitemsSource extends AirbyteSourceBase<AzureWorkitemsConfig> {
  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(
    config: AzureWorkitemsConfig
  ): Promise<[boolean, VError]> {
    try {
      const azureActiveDirectory = await AzureWorkitems.instance(config);
      await azureActiveDirectory.checkConnection();
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: AzureWorkitemsConfig): AirbyteStreamBase[] {
    return [
      new Workitems(config, this.logger),
      new Users(config, this.logger),
      new Iterations(config, this.logger),
      new Boards(config, this.logger),
    ];
  }
}
