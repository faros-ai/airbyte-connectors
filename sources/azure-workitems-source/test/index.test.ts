import {
  AirbyteLogger,
  AirbyteLogLevel,
  AirbyteSpec,
  SyncMode,
} from 'faros-airbyte-cdk';
import fs from 'fs-extra';
import {VError} from 'verror';

import {AzureWorkitems} from '../src/azure-workitems';
import * as sut from '../src/index';

const azureWorkitem = AzureWorkitems.instance;

jest.mock('axios');

describe('index', () => {
  const logger = new AirbyteLogger(
    // Shush messages in tests, unless in debug
    process.env.LOG_LEVEL === 'debug'
      ? AirbyteLogLevel.DEBUG
      : AirbyteLogLevel.INFO
  );

  beforeEach(() => {
    AzureWorkitems.instance = azureWorkitem;
  });

  function readResourceFile(fileName: string): any {
    return JSON.parse(fs.readFileSync(`resources/${fileName}`, 'utf8'));
  }

  function readTestResourceFile(fileName: string): any {
    return JSON.parse(fs.readFileSync(`test_files/${fileName}`, 'utf8'));
  }

  test('spec', async () => {
    const source = new sut.AzureWorkitemsSource(logger);
    await expect(source.spec()).resolves.toStrictEqual(
      new AirbyteSpec(readResourceFile('spec.json'))
    );
  });

  test('check connection - no access token', async () => {
    const source = new sut.AzureWorkitemsSource(logger);
    await expect(
      source.checkConnection({
        access_token: '',
        organization: 'organization',
        project: 'project',
      } as any)
    ).resolves.toStrictEqual([
      false,
      new VError('access_token must not be an empty string'),
    ]);
  });

  test('streams - users, use full_refresh sync mode', async () => {
    const fnUsersFunc = jest.fn();

    AzureWorkitems.instance = jest.fn().mockImplementation(() => {
      const usersResource: any[] = readTestResourceFile('users.json');
      return new AzureWorkitems(null, {
        get: fnUsersFunc.mockResolvedValue({
          data: {value: usersResource},
        }),
      } as any);
    });
    const source = new sut.AzureWorkitemsSource(logger);
    const streams = source.streams({} as any);

    const usersStream = streams[1];
    const userIter = usersStream.readRecords(SyncMode.FULL_REFRESH);
    const users = [];
    for await (const user of userIter) {
      users.push(user);
    }

    expect(fnUsersFunc).toHaveBeenCalledTimes(1);
    expect(users).toStrictEqual(readTestResourceFile('users.json'));
  });

  test('streams - iterations, use full_refresh sync mode', async () => {
    const fnIterationsFunc = jest.fn();

    AzureWorkitems.instance = jest.fn().mockImplementation(() => {
      const iterationsResource: any[] = readTestResourceFile('iterations.json');
      return new AzureWorkitems(
        {
          get: fnIterationsFunc.mockResolvedValue({
            data: {value: iterationsResource},
          }),
        } as any,
        null
      );
    });
    const source = new sut.AzureWorkitemsSource(logger);
    const streams = source.streams({} as any);

    const iterationsStream = streams[2];
    const iterationsIter = iterationsStream.readRecords(SyncMode.FULL_REFRESH);
    const iterations = [];
    for await (const iteration of iterationsIter) {
      iterations.push(iteration);
    }

    expect(fnIterationsFunc).toHaveBeenCalledTimes(3);
    expect(iterations).toStrictEqual(readTestResourceFile('iterations.json'));
  });

  test('streams - boards, use full_refresh sync mode', async () => {
    const fnBoardsFunc = jest.fn();

    AzureWorkitems.instance = jest.fn().mockImplementation(() => {
      const boardsResource: any[] = readTestResourceFile('boards.json');
      return new AzureWorkitems(
        {
          get: fnBoardsFunc.mockResolvedValue({
            data: {value: boardsResource},
          }),
        } as any,
        null
      );
    });
    const source = new sut.AzureWorkitemsSource(logger);
    const streams = source.streams({} as any);

    const usersStream = streams[3];
    const boardIter = usersStream.readRecords(SyncMode.FULL_REFRESH);
    const boards = [];
    for await (const board of boardIter) {
      boards.push(board);
    }

    expect(fnBoardsFunc).toHaveBeenCalledTimes(1);
    expect(boards).toStrictEqual(readTestResourceFile('boards.json'));
  });
});
