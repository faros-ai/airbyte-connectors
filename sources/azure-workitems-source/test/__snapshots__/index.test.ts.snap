// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index streams - iterations, use full_refresh sync mode 1`] = `
[
  {
    "attributes": undefined,
    "hasChildren": true,
    "id": 1,
    "identifier": "dd643523-4d15-4106-a6a4-e70d39a09446",
    "name": "Final Iteration",
    "path": "\\fabrikam",
    "url": "https://dev.azure.com/fabrikam/6ce954b1-ce1f-45d1-b94d-e6bf2464ba2c/_apis/wit/classificationNodes/Iterations",
  },
  {
    "attributes": {
      "finishDate": "2024-12-31T00:00:00Z",
      "startDate": "2024-12-01T00:00:00Z",
    },
    "hasChildren": false,
    "id": 2,
    "identifier": "6b5153a0-76f7-4aad-bd60-7dbfe3c2947d",
    "name": "Release 3",
    "path": "\\fabrikam\\Release 3",
    "structureType": "iteration",
    "url": "https://dev.azure.com/fabrikam/6ce954b1-ce1f-45d1-b94d-e6bf2464ba2c/_apis/wit/classificationNodes/Iterations/Release%203",
  },
  {
    "hasChildren": true,
    "id": 3,
    "identifier": "395a6b4f-6488-41ee-a06a-1514405ab6f0",
    "name": "Release 1",
    "path": "\\fabrikam\\Release 1",
    "structureType": "iteration",
    "url": "https://dev.azure.com/fabrikam/6ce954b1-ce1f-45d1-b94d-e6bf2464ba2c/_apis/wit/classificationNodes/Iterations/Release%201",
  },
  {
    "hasChildren": false,
    "id": 34,
    "identifier": "63d12e9d-37fd-48af-80a5-b93095705806",
    "name": "Sprint 4",
    "path": "\\fabrikam\\Release 1\\Sprint 4",
    "structureType": "iteration",
    "url": "https://dev.azure.com/fabrikam/6ce954b1-ce1f-45d1-b94d-e6bf2464ba2c/_apis/wit/classificationNodes/Iterations/Release%201/Sprint%204",
  },
  {
    "attributes": {
      "finishDate": "2024-09-07T00:00:00Z",
      "startDate": "2024-08-17T00:00:00Z",
    },
    "hasChildren": false,
    "id": 32,
    "identifier": "c9206006-4362-4b16-9205-1b94d32a6fb2",
    "name": "Sprint 2",
    "path": "\\fabrikam\\Release 1\\Sprint 2",
    "structureType": "iteration",
    "url": "https://dev.azure.com/fabrikam/6ce954b1-ce1f-45d1-b94d-e6bf2464ba2c/_apis/wit/classificationNodes/Iterations/Release%201/Sprint%202",
  },
  {
    "hasChildren": false,
    "id": 4,
    "identifier": "8c80c27e-8afb-4315-9057-686a1b862ed5",
    "name": "Release 2",
    "path": "\\fabrikam\\Release 2",
    "structureType": "iteration",
    "url": "https://dev.azure.com/fabrikam/6ce954b1-ce1f-45d1-b94d-e6bf2464ba2c/_apis/wit/classificationNodes/Iterations/Release%202",
  },
]
`;

exports[`index streams - workitems, use full_refresh sync mode 1`] = `
[
  {
    "additionalFields": [
      {
        "name": "Area Path",
        "value": "Fabrikam-Fiber-Git",
      },
      {
        "name": "Effort",
        "value": 8,
      },
    ],
    "fields": {
      "Faros": {
        "WorkItemStateCategory": {
          "category": "Proposed",
          "name": "New",
        },
      },
      "Microsoft.VSTS.Scheduling.Effort": 8,
      "System.AreaPath": "Fabrikam-Fiber-Git",
      "System.ChangedBy": {
        "_links": {
          "avatar": {
            "href": "https://dev.azure.com/mseng/_apis/GraphProfile/MemberAvatars/aad.YTkzODFkODYtNTYxYS03ZDdiLWJjM2QtZDUzMjllMjM5OTAz",
          },
        },
        "descriptor": "aad.YTkzODFkODYtNTYxYS03ZDdiLWJjM2QtZDUzMjllMjM5OTAz",
        "displayName": "Jamal Hartnett",
        "id": "d291b0c4-a05c-4ea6-8df1-4b41d5f39eff",
        "imageUrl": "https://dev.azure.com/fabrikam/_api/_common/identityImage?id=d291b0c4-a05c-4ea6-8df1-4b41d5f39eff",
        "uniqueName": "fabrikamfiber4@hotmail.com",
        "url": "https://vssps.dev.azure.com/fabrikam/_apis/Identities/d291b0c4-a05c-4ea6-8df1-4b41d5f39eff",
      },
      "System.ChangedDate": "2014-12-29T20:49:20.77Z",
      "System.CreatedBy": {
        "_links": {
          "avatar": {
            "href": "https://dev.azure.com/mseng/_apis/GraphProfile/MemberAvatars/aad.YTkzODFkODYtNTYxYS03ZDdiLWJjM2QtZDUzMjllMjM5OTAz",
          },
        },
        "descriptor": "aad.YTkzODFkODYtNTYxYS03ZDdiLWJjM2QtZDUzMjllMjM5OTAz",
        "displayName": "Jamal Hartnett",
        "id": "d291b0c4-a05c-4ea6-8df1-4b41d5f39eff",
        "imageUrl": "https://dev.azure.com/fabrikam/_api/_common/identityImage?id=d291b0c4-a05c-4ea6-8df1-4b41d5f39eff",
        "uniqueName": "fabrikamfiber4@hotmail.com",
        "url": "https://vssps.dev.azure.com/fabrikam/_apis/Identities/d291b0c4-a05c-4ea6-8df1-4b41d5f39eff",
      },
      "System.CreatedDate": "2014-12-29T20:49:20.77Z",
      "System.Description": "Our authorization logic needs to allow for users with Microsoft accounts (formerly Live Ids) - http://msdn.microsoft.com/en-us/library/live/hh826547.aspx",
      "System.IterationPath": "Fabrikam-Fiber-Git",
      "System.Reason": "New backlog item",
      "System.State": "New",
      "System.TeamProject": "Fabrikam-Fiber-Git",
      "System.Title": "Customer can sign in using their Microsoft Account",
      "System.WorkItemType": "Product Backlog Item",
      "WEF_6CB513B6E70E43499D9FC94E5BBFB784_Kanban.Column": "New",
    },
    "id": 297,
    "projectId": "123",
    "rev": 1,
    "revisions": {
      "assignees": [
        {
          "assignee": {
            "_links": {
              "avatar": {
                "href": "https://dev.azure.com/mseng/_apis/GraphProfile/MemberAvatars/aad.YTkzODFkODYtNTYxYS03ZDdiLWJjM2QtZDUzMjllMjM5OTAz",
              },
            },
            "descriptor": "aad.YTkzODFkODYtNTYxYS03ZDdiLWJjM2QtZDUzMjllMjM5OTAz",
            "displayName": "Jamal Hartnett",
            "id": "d291b0c4-a05c-4ea6-8df1-4b41d5f39eff",
            "imageUrl": "https://dev.azure.com/fabrikam/_api/_common/identityImage?id=d291b0c4-a05c-4ea6-8df1-4b41d5f39eff",
            "uniqueName": "fabrikamfiber4@hotmail.com",
            "url": "https://vssps.dev.azure.com/fabrikam/_apis/Identities/d291b0c4-a05c-4ea6-8df1-4b41d5f39eff",
          },
          "changedDate": "2017-09-04T02:28:56.253Z",
        },
      ],
      "iterations": [
        {
          "addedAt": "2017-09-04T02:08:16.6Z",
          "iteration": 3,
          "removedAt": "2017-09-04T02:36:18.75Z",
        },
        {
          "addedAt": "2017-09-04T02:36:18.75Z",
          "iteration": 4,
          "removedAt": null,
        },
      ],
      "states": [
        {
          "changedDate": "2017-09-04T02:08:16.6Z",
          "state": {
            "category": "Proposed",
            "name": "New",
          },
        },
        {
          "changedDate": "2017-09-04T02:36:18.75Z",
          "state": {
            "category": "InProgress",
            "name": "Active",
          },
        },
        {
          "changedDate": "2017-09-04T02:36:41.253Z",
          "state": {
            "category": "Resolved",
            "name": "Resolved",
          },
        },
      ],
    },
    "url": "https://dev.azure.com/fabrikam/_apis/wit/workItems/297",
  },
  {
    "additionalFields": [
      {
        "name": "Area Path",
        "value": "Fabrikam-Fiber-Git\\Website",
      },
      {
        "name": "Remaining Work",
        "value": 4,
      },
    ],
    "fields": {
      "Faros": {
        "WorkItemStateCategory": {
          "category": "Proposed",
          "name": "To Do",
        },
      },
      "Microsoft.VSTS.Scheduling.RemainingWork": 4,
      "System.AreaPath": "Fabrikam-Fiber-Git\\Website",
      "System.AssignedTo": {
        "_links": {
          "avatar": {
            "href": "https://dev.azure.com/mseng/_apis/GraphProfile/MemberAvatars/aad.YTkzODFkODYtNTYxYS03ZDdiLWJjM2QtZDUzMjllMjM5OTAz",
          },
        },
        "descriptor": "aad.YTkzODFkODYtNTYxYS03ZDdiLWJjM2QtZDUzMjllMjM5OTAz",
        "displayName": "Jamal Hartnett",
        "id": "d291b0c4-a05c-4ea6-8df1-4b41d5f39eff",
        "imageUrl": "https://dev.azure.com/fabrikam/_api/_common/identityImage?id=d291b0c4-a05c-4ea6-8df1-4b41d5f39eff",
        "uniqueName": "fabrikamfiber4@hotmail.com",
        "url": "https://vssps.dev.azure.com/fabrikam/_apis/Identities/d291b0c4-a05c-4ea6-8df1-4b41d5f39eff",
      },
      "System.ChangedBy": {
        "_links": {
          "avatar": {
            "href": "https://dev.azure.com/mseng/_apis/GraphProfile/MemberAvatars/aad.YTkzODFkODYtNTYxYS03ZDdiLWJjM2QtZDUzMjllMjM5OTAz",
          },
        },
        "descriptor": "aad.YTkzODFkODYtNTYxYS03ZDdiLWJjM2QtZDUzMjllMjM5OTAz",
        "displayName": "Jamal Hartnett",
        "id": "d291b0c4-a05c-4ea6-8df1-4b41d5f39eff",
        "imageUrl": "https://dev.azure.com/fabrikam/_api/_common/identityImage?id=d291b0c4-a05c-4ea6-8df1-4b41d5f39eff",
        "uniqueName": "fabrikamfiber4@hotmail.com",
        "url": "https://vssps.dev.azure.com/fabrikam/_apis/Identities/d291b0c4-a05c-4ea6-8df1-4b41d5f39eff",
      },
      "System.ChangedDate": "2014-12-29T20:49:28.74Z",
      "System.CreatedBy": {
        "_links": {
          "avatar": {
            "href": "https://dev.azure.com/mseng/_apis/GraphProfile/MemberAvatars/aad.YTkzODFkODYtNTYxYS03ZDdiLWJjM2QtZDUzMjllMjM5OTAz",
          },
        },
        "descriptor": "aad.YTkzODFkODYtNTYxYS03ZDdiLWJjM2QtZDUzMjllMjM5OTAz",
        "displayName": "Jamal Hartnett",
        "id": "d291b0c4-a05c-4ea6-8df1-4b41d5f39eff",
        "imageUrl": "https://dev.azure.com/fabrikam/_api/_common/identityImage?id=d291b0c4-a05c-4ea6-8df1-4b41d5f39eff",
        "uniqueName": "fabrikamfiber4@hotmail.com",
        "url": "https://vssps.dev.azure.com/fabrikam/_apis/Identities/d291b0c4-a05c-4ea6-8df1-4b41d5f39eff",
      },
      "System.CreatedDate": "2014-12-29T20:49:21.617Z",
      "System.Description": "Follow the code samples from MSDN",
      "System.IterationPath": "Fabrikam-Fiber-Git",
      "System.Reason": "New task",
      "System.State": "To Do",
      "System.Tags": "Tag1; Tag2",
      "System.TeamProject": "Fabrikam-Fiber-Git",
      "System.Title": "JavaScript implementation for Microsoft Account",
      "System.WorkItemType": "Task",
    },
    "id": 299,
    "projectId": "123",
    "rev": 7,
    "revisions": {
      "assignees": [
        {
          "assignee": {
            "_links": {
              "avatar": {
                "href": "https://dev.azure.com/mseng/_apis/GraphProfile/MemberAvatars/aad.YTkzODFkODYtNTYxYS03ZDdiLWJjM2QtZDUzMjllMjM5OTAz",
              },
            },
            "descriptor": "aad.YTkzODFkODYtNTYxYS03ZDdiLWJjM2QtZDUzMjllMjM5OTAz",
            "displayName": "Jamal Hartnett",
            "id": "d291b0c4-a05c-4ea6-8df1-4b41d5f39eff",
            "imageUrl": "https://dev.azure.com/fabrikam/_api/_common/identityImage?id=d291b0c4-a05c-4ea6-8df1-4b41d5f39eff",
            "uniqueName": "fabrikamfiber4@hotmail.com",
            "url": "https://vssps.dev.azure.com/fabrikam/_apis/Identities/d291b0c4-a05c-4ea6-8df1-4b41d5f39eff",
          },
          "changedDate": "2017-09-04T02:28:56.253Z",
        },
      ],
      "iterations": [
        {
          "addedAt": "2017-09-04T02:08:16.6Z",
          "iteration": 3,
          "removedAt": "2017-09-04T02:36:18.75Z",
        },
        {
          "addedAt": "2017-09-04T02:36:18.75Z",
          "iteration": 4,
          "removedAt": null,
        },
      ],
      "states": [
        {
          "changedDate": "2017-09-04T02:08:16.6Z",
          "state": {
            "category": "Proposed",
            "name": "New",
          },
        },
        {
          "changedDate": "2017-09-04T02:36:18.75Z",
          "state": {
            "category": "InProgress",
            "name": "Active",
          },
        },
        {
          "changedDate": "2017-09-04T02:36:41.253Z",
          "state": {
            "category": "Resolved",
            "name": "Resolved",
          },
        },
      ],
    },
    "url": "https://dev.azure.com/fabrikam/_apis/wit/workItems/299",
  },
  {
    "additionalFields": [
      {
        "name": "Area Path",
        "value": "Fabrikam-Fiber-Git",
      },
      {
        "name": "Remaining Work",
        "value": 3,
      },
      {
        "name": "Test Name",
        "value": "Zephyr",
      },
    ],
    "fields": {
      "Custom.TestName": "Zephyr",
      "Faros": {
        "WorkItemStateCategory": {
          "category": "Proposed",
          "name": "To Do",
        },
      },
      "Microsoft.VSTS.Scheduling.RemainingWork": 3,
      "System.AreaPath": "Fabrikam-Fiber-Git",
      "System.ChangedBy": {
        "_links": {
          "avatar": {
            "href": "https://dev.azure.com/mseng/_apis/GraphProfile/MemberAvatars/aad.YTkzODFkODYtNTYxYS03ZDdiLWJjM2QtZDUzMjllMjM5OTAz",
          },
        },
        "descriptor": "aad.YTkzODFkODYtNTYxYS03ZDdiLWJjM2QtZDUzMjllMjM5OTAz",
        "displayName": "Jamal Hartnett",
        "id": "d291b0c4-a05c-4ea6-8df1-4b41d5f39eff",
        "imageUrl": "https://dev.azure.com/fabrikam/_api/_common/identityImage?id=d291b0c4-a05c-4ea6-8df1-4b41d5f39eff",
        "uniqueName": "fabrikamfiber4@hotmail.com",
        "url": "https://vssps.dev.azure.com/fabrikam/_apis/Identities/d291b0c4-a05c-4ea6-8df1-4b41d5f39eff",
      },
      "System.ChangedDate": "2014-12-29T20:49:22.103Z",
      "System.CreatedBy": {
        "_links": {
          "avatar": {
            "href": "https://dev.azure.com/mseng/_apis/GraphProfile/MemberAvatars/aad.YTkzODFkODYtNTYxYS03ZDdiLWJjM2QtZDUzMjllMjM5OTAz",
          },
        },
        "descriptor": "aad.YTkzODFkODYtNTYxYS03ZDdiLWJjM2QtZDUzMjllMjM5OTAz",
        "displayName": "Jamal Hartnett",
        "id": "d291b0c4-a05c-4ea6-8df1-4b41d5f39eff",
        "imageUrl": "https://dev.azure.com/fabrikam/_api/_common/identityImage?id=d291b0c4-a05c-4ea6-8df1-4b41d5f39eff",
        "uniqueName": "fabrikamfiber4@hotmail.com",
        "url": "https://vssps.dev.azure.com/fabrikam/_apis/Identities/d291b0c4-a05c-4ea6-8df1-4b41d5f39eff",
      },
      "System.CreatedDate": "2014-12-29T20:49:22.103Z",
      "System.Description": "We need to ensure we have coverage to prevent regressions",
      "System.IterationPath": "Fabrikam-Fiber-Git",
      "System.Reason": "New task",
      "System.State": "To Do",
      "System.TeamProject": "Fabrikam-Fiber-Git",
      "System.Title": "Unit Testing for MSA login",
      "System.WorkItemType": "Task",
    },
    "id": 300,
    "projectId": "123",
    "rev": 1,
    "revisions": {
      "assignees": [
        {
          "assignee": {
            "_links": {
              "avatar": {
                "href": "https://dev.azure.com/mseng/_apis/GraphProfile/MemberAvatars/aad.YTkzODFkODYtNTYxYS03ZDdiLWJjM2QtZDUzMjllMjM5OTAz",
              },
            },
            "descriptor": "aad.YTkzODFkODYtNTYxYS03ZDdiLWJjM2QtZDUzMjllMjM5OTAz",
            "displayName": "Jamal Hartnett",
            "id": "d291b0c4-a05c-4ea6-8df1-4b41d5f39eff",
            "imageUrl": "https://dev.azure.com/fabrikam/_api/_common/identityImage?id=d291b0c4-a05c-4ea6-8df1-4b41d5f39eff",
            "uniqueName": "fabrikamfiber4@hotmail.com",
            "url": "https://vssps.dev.azure.com/fabrikam/_apis/Identities/d291b0c4-a05c-4ea6-8df1-4b41d5f39eff",
          },
          "changedDate": "2017-09-04T02:28:56.253Z",
        },
      ],
      "iterations": [
        {
          "addedAt": "2017-09-04T02:08:16.6Z",
          "iteration": 3,
          "removedAt": "2017-09-04T02:36:18.75Z",
        },
        {
          "addedAt": "2017-09-04T02:36:18.75Z",
          "iteration": 4,
          "removedAt": null,
        },
      ],
      "states": [
        {
          "changedDate": "2017-09-04T02:08:16.6Z",
          "state": {
            "category": "Proposed",
            "name": "New",
          },
        },
        {
          "changedDate": "2017-09-04T02:36:18.75Z",
          "state": {
            "category": "InProgress",
            "name": "Active",
          },
        },
        {
          "changedDate": "2017-09-04T02:36:41.253Z",
          "state": {
            "category": "Resolved",
            "name": "Resolved",
          },
        },
      ],
    },
    "url": "https://dev.azure.com/fabrikam/_apis/wit/workItems/300",
  },
]
`;
